src/compute/src/render/flat_map.rs

```diff
@@ -106,3 +117,76 @@ where
         CollectionBundle::from_collections(ok_collection, err_collection)
     }
 }
+
+use crate::render::DataflowError;
+use mz_expr::MfpPlan;
+use mz_repr::{Diff, Row, Timestamp};
+use timely::progress::Antichain;
+
+/// Drains a list of extensions to `input_row` through a supplied `MfpPlan` and into output buffers.
+fn drain_through_mfp<T>(
+    input_row: &Row,
+    input_time: &mut T,
+    input_diff: &Diff,
+    datum_vec: &mut DatumVec,
+    extensions: &[(Row, Diff)],
+    mfp_plan: &MfpPlan,
+    until: &Antichain<Timestamp>,
+    oks_output: &mut Vec<(Row, T, Diff)>,
+    err_output: &mut Vec<(DataflowError, T, Diff)>,
+) where
+    T: crate::render::RenderTimestamp,
+{
+    let temp_storage = RowArena::new();
+    let binding = SharedRow::get();
+    let mut row_builder = binding.borrow_mut();
+
+    let mut datums_local = datum_vec.borrow_with(input_row);
+    let datums_len = datums_local.len();
```

Comment:
Observing that this borrows `input_row` for every call to `drain_through_mfp`, which is probably OK based on how (in)frequently it'll be called.

Reply:
Yes I'll add a comment to that effect. It seems to be the .. only/best way to make the lifetimes work out, that I could determine.